Add PayatTable features and Pre-auth features

metinavc1 opened the pull request.

Commit:
Add PayatTable features and Pre-auth features

gouline commented on Library/Model/SPIPayAtTable.h:
```diff
@@ -100,15 +100,16 @@ typedef NS_ENUM(NSUInteger, SPIPaymentType) {
-@property (nonatomic, readonly, copy) NSString *labelOperatorId;
-@property (nonatomic, readonly, copy) NSString *labelTableId;
-@property (nonatomic, readonly, copy) NSArray<NSString *> *allowedOperatorIds;
+@property (nonatomic) NSString *labelPayButton;
```
Reason for change?

metinavc1 replied:
They couldn't be set a value when they were read-only

gouline replied:
I was more talking about the removal of `copy`, this will impact the behaviour.

metinavc1 replied:
ok I got it. Should it be like that `@property (nonatomic, copy) NSString *labelPayButton;`?

chrishulbert replied:
Copy is usually a safe bet for things like strings / arrays where performance isn't a massive deal.

metinavc1 replied:
Added like that @property (nonatomic, copy) NSString *labelPayButton; @chrishulbert

Commit:
Fix handlePreathMessage method

Commit:
Fix PaymentType lowercase string

Commit:
Fix bugs SCLI-34

SCLI-34 Pay at Table - CARD is displayed in the receipt instead of CASH for cash payments

Commit:
Added SCLI-13

SCLI-13 API available to disable pay at table

chrishulbert commented:
Hi, I’ll be glad to have a look at this if my contract gets renewed in the new year :) 

Sent from my iPhone

> On 28 Dec 2018, at 10:02 am, Raymond Ko <notifications@github.com> wrote:
> 
> @RaymondK-AP requested your review on: #25 Add PayatTable features and Pre-auth features.
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub, or mute the thread.

Commit:
SCLI-22 Allow surcharging from POS for Pre-Auth Completion

Commit:
Adding get Surcharge Amount for Pre-Auth Completion

chrishulbert commented on Library/Model/SPIPayAtTable.m:
```diff
@@ -19,57 +19,64 @@ @implementation SPIBillStatusResponse
     NSMutableDictionary *data = [[NSMutableDictionary alloc] init];
-    [data setObject:[NSNumber numberWithInteger:BillRetrievalResultSuccess] forKey:@"success"];
+    BOOL success = _result == BillRetrievalResultSuccess;
+    [data setValue:[NSNumber numberWithBool:success] forKey:@"success"];
```
IIRC, this is a shortcut for nsnumber:
```suggestion
    [data setValue:@success forKey:@"success"];
```

metinavc1 replied:
it gives the error "Unexpected ''@" in program " when I change it like your suggestion @chrishulbert

chrishulbert replied:
My ObjC is rusty... perhaps it's `@(success)`. Nevermind, it's just a style suggestion to keep in mind :)

metinavc1 replied:
Done

chrishulbert commented on Library/Model/SPIPayAtTable.m:
```diff
@@ -19,57 +19,64 @@ @implementation SPIBillStatusResponse
-        [data setObject:@"bill_total_amount" forKey:[NSNumber numberWithInteger:_totalAmount]];
-        [data setObject:@"bill_outstanding_amount" forKey:[NSNumber numberWithInteger:_outstandingAmount]];
-        [data setObject:@"bill_payment_history" forKey:[self getBillPaymentHistory]];
+        [data setValue:[NSNumber numberWithInteger:_totalAmount] forKey:@"bill_total_amount"];
```
Should work if you want to remember for next time, no big issue if you dont do this for this version: ```suggestion
        [data setValue:@_totalAmount forKey:@"bill_total_amount"];
```

metinavc1 replied:
it gives the error "Unexpected ''@" in program " when I change it like your suggestion @chrishulbert

metinavc1 replied:
Done

chrishulbert commented on Library/Model/SPIPayAtTable.m:
```diff
@@ -112,35 +119,45 @@ - (instancetype)initWithMessage:(SPIMessage *)message {
     }
-    NSDictionary<NSString *,NSObject *> *data = (NSDictionary *)[message.data valueForKey:@"payment_details"];
     
+    NSDictionary<NSString *,NSObject *> *data = (NSDictionary *)[message.data valueForKey:@"payment_details"];
```
```suggestion
    NSDictionary<NSString *,NSObject *> *data = (NSDictionary *)message.data[@"payment_details"];
```
Up to you if you want to use this shortcut in future

metinavc1 replied:
ok i got it

chrishulbert commented on Library/Model/SPIPayAtTable.m:
```diff
@@ -149,17 +166,17 @@ @implementation SPIPayAtTableConfig
-    [data setValue:[NSNumber numberWithBool:_labelPayButton] forKey:@"pay_button_label"];
-    [data setValue:[NSNumber numberWithBool:_labelOperatorId] forKey:@"operator_id_label"];
-    [data setValue:[NSNumber numberWithBool:_labelTableId] forKey:@"table_id_label"];
+    [data setValue:[NSString stringWithFormat:@"%@", _labelPayButton] forKey:@"pay_button_label"];
```
This (and the next 2 lines) looks peculiar. Are we serialising the descriptions of buttons into JSON? That doesn't seem right. Am i missing something here? Should it rather be something like `setValue:_labelPayButton.text` ?

metinavc1 replied:
I checked it that doesn't work.

chrishulbert replied:
So i’ve looked in the header, labelPayButton isn’t actually a UIButton which i assumed, it’s just a NSString
In which case there’s no point doing the stringWithFormat, you can just set it as the dictionary value directly
but there’s no harm either, it’s just an odd thing, so i’m happy to approve

chrishulbert commented on Library/Model/SPIPayAtTable.m:
```diff
@@ -184,6 +201,18 @@ @implementation SPIPayAtTable
 
 - (instancetype)initWithClient:(SPIClient *)spi {
     _spi = spi;
+    _config = [SPIPayAtTableConfig alloc];
```
please call `init` as well :)

metinavc1 replied:
Done

chrishulbert commented on Library/Model/SPIPreAuth.h:
```diff
@@ -7,11 +7,186 @@
 
 @class SPIClient;
 
+typedef void (^SPICompletionTxResult)(SPIInitiateTxResult *result);
```
Great to see a block typedef :)

metinavc1 replied:
👍 thanks :)

chrishulbert commented on Library/Model/SPIPreAuth.h:
```diff
@@ -7,11 +7,186 @@
+@interface SPIAccountVerifyResponse : NSObject
+
+@property (nonatomic, readonly, copy) NSString *posRefId;
+@property (nonatomic, readonly, copy) SPIPurchaseResponse *details;
```
Normally you just use 'copy' for strings/arrays/dictionaries - you can certainly use it for your own types like this but be careful that the type has NSCopying implemented that's all :)

metinavc1 replied:
ok i understand :)

chrishulbert commented on Library/SPIClient.m:
```diff
@@ -103,6 +103,12 @@ - (SPIPayAtTable *)enablePayAtTable {
 
+- (SPIPayAtTable *)disablePayAtTable {
+    _spiPat = [[SPIPayAtTable alloc] initWithClient:self];
+    _spiPat.config.payAtTableEnabled = true;
```
should this be false to disable it as per the method name?

metinavc1 replied:
yes right thanks :)

chrishulbert commented:
Ok my review is complete, if you can just work through my comments, let me know how you go, and i'll approve it, nothing major, mostly just things i think might be worth double checking :)

Commit:
Fix the coding aspect according to code review

chrishulbert reviewed (changes requested):
Just a couple little changes cheers 🍻

Commit:
Fix coding according to code review

Commit:
Merge branch 'feature/payattable-preauth-trans' of https://github.com/AssemblyPayments/spi-client-ios into feature/payattable-preauth-trans

Commit:
Merge branch 'master' into feature/payattable-preauth-trans

Commit:
Merge branch 'feature/payattable-preauth-trans' of https://github.com/AssemblyPayments/spi-client-ios into feature/payattable-preauth-trans

Commit:
Merge branch 'feature/payattable-preauth-trans' of https://github.com/AssemblyPayments/spi-client-ios into feature/payattable-preauth-trans

Commit:
Merge branch 'master' into feature/payattable-preauth-trans
